### ISSUE TITLE
Voting On Features, Stories

### ISSUE THREAD

**Issue by tansaku** [AgileVentures/AgileVentures#8]
do we need more rigour on the voting of stories, and criteria for when we can vote on a story and whether stories have to be voted on before work can start?
- (michael) voting can help bring out different ideas - possibly help with prediction
- (tony) concerned that voting might put people off
- (raoul) not sure how this works for waffle, and not so many people on project, and worried about slowing things down

**Comment by tochman** [comment]
This feels like a WSO matter or it this question more of a general character? 
About voting/assessing stories: I would say a return to a more rigorous approach to the voting and assessment process would be a good thing for many projects, especially WSO. The way I see it the voting process is an fundamental part of the agile process. It gives the dev team an opportunity to discuss the feature beyond the often short and cryptic description in the feature request, reach a consensus on the scope of the feature and the definition of 'done' and also make a prioritization of the feature in relation to other requirements. 

Also, in the case of WSO, the feature requester can often be a person not directly involved in the development process (we don't have a Client in that project, right? A request can be made by any community member). This can mean that this person will not be able to take part in the day to day discussions of the dev team and his or hers input will be lost in the general chatter. Maintaining a discipline of following up all requests/tickets and sticking to the agile process feels important. :-)

**Comment by tansaku** [comment]
This question was originally raised in the context of WSO, but I think it's relevant to all projects, and I often field questions about it for other projects, so this seemed a better place to discuss.

I've been a big proponent of voting on stories over the last few years, however I think the project gets the most benefit from it when you have a committed, regularly available team.  It can be also useful for a distributed volunteer team, but I start to suspect that the benefits are more for the learning developers than they are for the project per se.  I've put countless hours into voting on stories over the last few years, and I've seen a lot of benefit for individuals.  For projects?  I think it depends on the project.

Regarding your second point, maintaining the discipline of following up all requests/tickets and using the agile process; there are clearly two very important things, but I think they are two separate issues from voting on stories.  The Agile process (as I understand it) doesn't specify any individual workflow element (e.g. TDD, Story Voting, ticketing per se, or any tool).  From my point of view the key thing that makes a process Agile is that one plans and reflects on the result of those plans with some regular frequency.

Story voting does not happen in many open source projects and it is possible that the requirement for voting on a story may serve as a barrier to entry on a project for some developers.  That's not to say that projects should or shouldn't have story voting, but I think it's something that each agile team needs to way up the pros and cons of for their particular project, e.g. 

Pros
-----

* creates a space for everyone to input on a story
* can help expose different assumptions between developers
* can help new developers assess which story to start work on
* ...

Cons
-----

* Takes time
* Unfamiliar process may intimidate newcomers
* Apparent requirement to be present in a synchronous standup may serve as a barrier to entry for some
*...

And perhaps we can add to both lists.  Each project needs to take an informed decision on whether they include voting in their process, and keep an open mind about whether to do more or less story voting.

We just did a round of story voting in WSO last week, but we'll be keeping a close watch over whether this is really something we need to be particularly rigorous about given the nature of the project.

**Comment by freeranger** [comment]
I don't think that voting on stories is incompatible with the casual nature of many open source projects, in the same way that having daily/weekly team meetings does not have to be incompatible with lone developers contributing when they can.

The key is to not make any of this mandatory - yes, a core team can meet as often as they like and vote on the stories they would like to work on in the upcoming "sprint", but that should not preclude contribution from johnny lone developer who happened to have some spare time and the right skills and submits a PR for an issue which is not necessarily a priority but is still a valid piece of work.
Maybe the team members get more out of it with greater participation, and joining in could be encouraged but not at the expense of turning away people who want to contribute in other ways.

I guess PM's need to strike a balance between trying to run a project like they would at work, where they have a team of paid developers with known skills who's job is to follow the process and do what they are told vs a loose collection of volunteer programmers of varying ability with differing amounts of time they can put in to these projects who have their own way of working.

The latter may not work in the "preferred" way but they can certainly add a lot of value even if they don't participate in meetings and voting etc.  For example, looking on github at the most active projects over the last month and top of the list is: https://github.com/dotnet/corefx which is Microsoft's core .NET framework - this has had over 8,500 commits from 233 developers and is a massive project in production with hundreds of thousands of users.  They don't vote on issues or enforce a priority on stories - they may prioritise the PR's they review but that doesn't preclude anyone working on any issue at any time, or submitting a PR as and when they feel like it, and it seems to work.

Whilst we don't currently have any projects on the scale of this one, that isn't to say that could never happen.  Trying to enforce workplace-like rigour in an open source project would be nigh on impossible at that sort of scale.

**Comment by tochman** [comment]
Understood @tansaku The big question is if we are using scrum or any other methods in our projects. For me, personally, what has always distinguished AV projects from other OO projects was that we follow the methods and workflows we are likely to use in an professional agile environment.We used that in Local Support, which was the first project I worked on in AV and that was what I brought to the WSO project when I started it back in 2014. It gave us a framework to fall back on when there was a lot of features and functions constantly added to the application and gave us a valuable support for team communication (the voting process being one of many techniques). My experience, contrary to how other may see it, tells me that deviations from that leads to low activity, a massive technical debt and subsequently abandoned projects. (If in doubt do have a look at Project Unify, which is my latest project on AV, and see the effects of not closely following the chosen path.)

I also don't buy the `barrier to entry` argument (surprise, surprise!). ;-) I do OO software dev on AV and all things related "not because they are easy, but because they are hard; because that goal will serve to organize and measure the best of our energies and skills, because that challenge is one that we are willing to accept, one we are unwilling to postpone..." ;-)

Peace and LLAP ;-)
